cirq/contrib/two_qubit_gates/math_utils.py

```diff
@@ -0,0 +1,292 @@
+from itertools import permutations
+from typing import Tuple, Union
+
+import numpy as np
+
+TWO_PI = np.pi * 2
+
+_RealArraylike = Union[np.ndarray, float]
+
+
+def _single_qubit_unitary(theta: _RealArraylike, phi_d: _RealArraylike,
+                          phi_o: _RealArraylike) -> np.ndarray:
+    """Single qubit unitary matrix.
+
+    Args:
+        theta: cos(theta) is magnitude of 00 matrix element. May be a scalar
+           or real ndarray (for broadcasting).
+        phi_d: exp(i phi_d) is the phase of 00 matrix element. May be a scalar
+           or real ndarray (for broadcasting).
+        phi_o: i exp(i phi_o) is the phase of 10 matrix element. May be a scalar
+           or real ndarray (for broadcasting).
+
+
+    Notes:
+        The output is vectorized with respect to the angles. I.e, if the angles
+        are (broadcastable) arraylike objects whose sum would have shape (...),
+        the output is an array of shape (...,2,2), where the final two indices
+        correspond to unitary matrices.
+    """
+
+    U00 = np.cos(theta) * np.exp(1j * np.asarray(phi_d))
+    U10 = 1j * np.sin(theta) * np.exp(1j * np.asarray(phi_o))
+
+    # This implementation is agnostic to the shapes of the angles, as long
+    # as they can be broadcast together.
+    Udiag = np.array([[U00, np.zeros_like(U00)],
+                      [np.zeros_like(U00), U00.conj()]])
+    Udiag = np.moveaxis(Udiag, [0, 1], [-2, -1])
+    Uoff = np.array([[np.zeros_like(U10), -U10.conj()],
+                     [U10, np.zeros_like(U10)]])
+    Uoff = np.moveaxis(Uoff, [0, 1], [-2, -1])
+    return Udiag + Uoff
+
+
+def random_qubit_unitary(number: int = 1,
+                         sample_phase: bool = False) -> np.ndarray:
+    """Random qubit unitary distributed over the Haar measure.
+
+    The implementation is vectorized for speed.
+
+    Args:
+        number: If not 1, an ndarray of shape (number,2,2) is returned.
+            Otherwise a single 2x2 matrix is returned.
+        sample_phase: (Default False) If True, a global phase is also sampled
+            randomly. This corresponds to sampling over U(2) instead of SU(2).
+    """
+    theta = np.arcsin(np.sqrt(np.random.rand(number)))
+    phi_d = np.random.rand(number) * TWO_PI
+    phi_o = np.random.rand(number) * TWO_PI
+
+    out = _single_qubit_unitary(theta, phi_d, phi_o)
+
+    if sample_phase:
+        # coverage: ignore
+        global_phase = np.exp(1j * TWO_PI * np.random.rand(number))
+        np.einsum('t,tab->tab', global_phase, out, out=out)
+    return out
+
+
+def vector_kron(first: np.ndarray, second: np.ndarray) -> np.ndarray:
```

Comment:
Can we have linalg/decompositions_test.py reference this so it isn't duplicated?

Reply:
This could potentially be merged into `cirq.kron` in the form of an `axes` parameter.